README.md
# Consul Bak [![Build Status](https://travis-ci.org/Tubular/consul-bak.png)](https://travis-ci.org/Tubular/consul-bak)

Backs up, restores, syncs and dumps KV pairs in a Consul cluster using the consul-api Go library.

## Changelog / Releases

[See the changelog here](CHANGELOG.md)


## Modes


### Backup 

This backups your Consul master's KV store into a file. You can run this via a crontab and sync 
it to S3 for example:

```
filename="$(date +%Y%m%d).txt"
/usr/local/bin/consul-bak backup --leader-only $filename
/usr/local/bin/aws s3 mv $filename s3://<my_bucket>/
```


### Restore

This restores a file created by consul-bak into the Consul master's KV store. Example:

```
/usr/local/bin/consul-bak restore $filename
```


### Syncgit

This synchronises a filesystem representation of the KV tree into a Consul master's KV store. The
filesystem representation is expected to be a git repository. This is especially useful if you want
a code-first approach to creating configuration in Consul. Example:

```
/usr/local/bin/consul-bak sync git@github.com:Tubular/consul-bak.git|path/to/root/of/kv/tree
```


### Dumptree

This dumps the values of the Consul master's KV as a tree onto the filesystem. If you want to start
using sync, you could use this to kickstart your project. Example:

```
/usr/local/bin/consul-bak dumptree /tmp/kv_tree
```


## Usage examples

```sh
Usage:
  consul-bak (backup|restore|aclbackup)
                [--leader-only]
                [--rpc-port RPCPORT]
                [--http-port HTTPPORT]
                [--address IP]
                [--include-prefix INPREFIX]...
                [--exclude-prefix EXPREFIX]...
                [--token TOKEN]
                [--no-prompt]
                <filename>
  consul-bak dumptree
                [--leader-only]
                [--rpc-port RPCPORT]
                [--http-port HTTPPORT]
                [--address IP]
                <pathname>
  consul-bak syncgit
                [--leader-only]
                [--rpc-port RPCPORT]
                [--http-port HTTPPORT]
                [--address IP]
                <git-url>
  consul-bak -h | --help
  consul-bak -v | --version

Options:
  -h, --help                         Show this screen.
  -v, --version                      Show version.
	--mode=MODE                        Set mode, can be one of backup,restore,syncgit,dumptree,aclbackup [default: backup]
  --leader-only                      Only run on consul leader.
  --rpc-port=RPCPORT                 RPC port [default: 8400].
  --http-port=HTTPPORT               HTTP endpoint port [default: 8500].
  --address=IP                       The HTTP endpoint of Consul [default: 127.0.0.1].
  --include-prefix=[INPREFIX]        Repeatable option for keys starting with prefix to include in the backup.
  --exclude-prefix=[EXPREFIX]        Repeatable option for keys starting with prefix to exclude from the backup.
  --token=TOKEN                      An ACL Token with proper permissions in Consul [default: ].
  --force                            Don't prompt, force overwrite.
```



## Development

```
# Somewhere in your GOPATH
git clone git@github.com:Tubular/consul-bak.git

# Get requirements
glide install

# Run
make run
```

## Release

To bump a version, change the `Version` constant in version.go and use git tag:

```
vi version.go && git commit -m "Bump version to #.#.#" && git tag -a "Released #.#.#"
```


*Code inspired by work done here: https://github.com/kailunshi/consul-backup*


CHANGELOG.md
## 0.1.0 [Release page](https://github.com/Tubular/consul-bak/releases/tag/0.1.0)
* Improved logging
* Added dumptree command that creates a filesystem representation of the Consul KV store
* Added syncgit commant that reads a git repo and writes to Consul's KV store
* Changed syntax of the utility to be more straightforward (BREAKING)

## 0.0.1 [Release page](https://github.com/Tubular/consul-bak/releases/tag/0.0.1)
* Initial release


tests/util_test.go
package main_test

import (
	consulbak "github.com/Tubular/consul-bak"
	"testing"
)

type testCasePair struct {
	value  []string
	prefix string
	expect bool
}

var tests = []testCasePair{
	{[]string{"foo", "baz"}, "foobar", true},
	{[]string{"bar"}, "foobar", false},
}

func TestStartsWith(t *testing.T) {
	for _, testCase := range tests {
		result := consulbak.StartsWith(testCase.value, testCase.prefix)
		if result != testCase.expect {
			t.Errorf("For: %s in %s expected %t, actual: %t", testCase.value, testCase.prefix, testCase.expect, result)
		}
	}
}


dump.go
package main

import (
	"fmt"
	"github.com/hashicorp/consul/api"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

func Dump(ipAddress string, token string, rootPath string) {

	var standardDirPerms os.FileMode = 0775
	logger.Infof("Starting dump from %s to %s.", ipAddress, rootPath)

	config := api.DefaultConfig()
	config.Address = ipAddress
	config.Token = token

	client, _ := api.NewClient(config)
	kv := client.KV()

	pairs, _, err := kv.List("/", nil)
	if err != nil {
		panic(err)
	}

	sort.Sort(ByCreateIndex(pairs))

	if len(pairs) == 0 {
		logger.Info("No KV pairs found. Nothing to do. Exiting.")
		os.Exit(0)
	} else {
		logger.Infof("Found %d KV pairs", len(pairs))
	}

	oldDir, _ := os.Getwd()
	defer os.Chdir(oldDir)

	absPath, pathErr := filepath.Abs(rootPath)
	logger.Debugf("Path to dump kv in is: %s", absPath)
	Check(pathErr)
	ensureErr := EnsureDir(absPath, standardDirPerms)
	Check(ensureErr)
	chdirErr := os.Chdir(absPath)
	Check(chdirErr)
	baseDir, _ := os.Getwd()
	logger.Debugf("Changed directory to: %s", baseDir)

	for _, element := range pairs {
		var path string
		path = fmt.Sprintf("%s/%s", baseDir, element.Key)
		logger.Infof("Writing %s", path)
		if strings.HasSuffix(path, "/") {
			err := EnsureDir(path, standardDirPerms)
			Check(err)
		} else {
			// ensure directory is present before creating file
			err := EnsureDir(filepath.Dir(path), standardDirPerms)
			Check(err)

			f, createErr := os.Create(path)
			Check(createErr)
			f.Write(element.Value)
		}
	}
}


sync.go
package main

import (
	"fmt"
	"github.com/hashicorp/consul/api"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
)

// Sync reads a repo specified by `gitString` and writes it to Consul at `ipAddress`
func Sync(ipAddress string, token string, gitString string) {
	logger.Infof("Got git string: %s", gitString)
	const cacheDir = "/var/cache/consul_backup/"

	parts := strings.Split(gitString, "|")
	gitURL := parts[0]
	kvRootPath := parts[1]
	logger.Debugf("URL: %s, KV path: %s", gitURL, kvRootPath)

	if !strings.HasSuffix(kvRootPath, "/") {
		kvRootPath = kvRootPath + "/"
	}

	var rootDir string = fmt.Sprintf("%s%s", cacheDir, kvRootPath)

	if !Which("git") {
		logger.Error("You must have git installed to use syncgit mode.")
		os.Exit(1)
	}

	oldDir, _ := os.Getwd()
	defer os.Chdir(oldDir)
	repoExists, existsErr := PathExists(fmt.Sprintf("%s.git", cacheDir))
	Check(existsErr)

	if !repoExists {
		logger.Infof("Repo at %s does not exist yet, creating.", cacheDir)
		var perm os.FileMode = 0775
		mkdirErr := os.MkdirAll(cacheDir, perm)
		Check(mkdirErr)
		GitBinaryCall(fmt.Sprintf("clone %s %s", gitURL, cacheDir))
		chdirErr := os.Chdir(cacheDir)
		Check(chdirErr)
		GitBinaryCall("config core.sparseCheckout true")
		f, createErr := os.Create(".git/info/sparse-checkout")
		Check(createErr)
		f.WriteString(kvRootPath)
	}
	changeDirErr := os.Chdir(rootDir)
	Check(changeDirErr)
	currentDir, _ := os.Getwd()

	logger.Info("Pulling latest info from git.")
	GitBinaryCall("checkout master")
	GitBinaryCall("remote update")
	GitBinaryCall("pull origin master")

	config := api.DefaultConfig()
	config.Address = ipAddress
	config.Token = token

	client, _ := api.NewClient(config)
	kv := client.KV()

	logger.Info("Creating a list of keys")
	fileList := []string{}
	walkErr := filepath.Walk(currentDir, func(path string, f os.FileInfo, err error) error {
		logger.Info(path)
		fileList = append(fileList, path)
		return nil
	})
	Check(walkErr)

	for _, fileName := range fileList {
		fi, statErr := os.Stat(fileName)
		Check(statErr)

		var key string
		var val []byte
		var readErr error

		switch mode := fi.Mode(); {
		case mode.IsDir():
			key = fmt.Sprintf("%s/", fileName)
		case mode.IsRegular():
			key = fileName
			val, readErr = ioutil.ReadFile(fileName)
			Check(readErr)
		}
		key = strings.Replace(key, rootDir, "", 1)
		if key != "" {
			p := &api.KVPair{Key: key, Value: val}
			_, putErr := kv.Put(p, nil)
			Check(putErr)
		}
	}
	logger.Infof("Wrote %d KV pairs to Consul at %s from Git repo %s", len(fileList), ipAddress, gitURL)
}


utils.go
package main

import (
	"errors"
	"fmt"
	"net"
	"os"
	"os/exec"
	"strings"
)

// EnsureDir creates a directory at `path` if it doesn't already exist
func EnsureDir(path string, perm os.FileMode) error {
	logger.Infof("Making sure a directory exists at: %s", path)
	fi, err := os.Stat(path)
	if err == nil {
		if fi.IsDir() {
			return nil
		} else {
			err := errors.New(fmt.Sprintf("Cannot create directory '%s': File exists.", path))
			logger.Fatal(err)
			return err
		}
	} else if os.IsNotExist(err) {
		// nothing found, create a directory
		mkdirErr := os.MkdirAll(path, perm)
		if mkdirErr != nil {
			logger.Errorf("Failed to create a directory at: %s (%s)", path, mkdirErr)
			return mkdirErr
		}
		return nil
	}
	logger.Fatal(err)
	return err
}

// PathExists verifies that the path at path exists
func PathExists(path string) (bool, error) {
	_, err := os.Stat(path)
	if err == nil {
		return true, nil
	}
	if os.IsNotExist(err) {
		return false, nil
	}
	return true, err
}

// Check is a utility function for checking error returns
func Check(e error) {
	if e != nil {
		panic(e)
	}
}

// StartsWith checks to verify if a string starts with a given substring
func StartsWith(list []string, elem string) bool {
	for _, t := range list {
		if strings.HasPrefix(elem, t) {
			return true
		}
	}
	return false
}

// CheckSocket checks to see if a socket `endpoint` is listening.
func CheckSocket(endpoint string) bool {
	logger.Debugf("Checking that socket %s is listening", endpoint)
	_, err := net.Dial("tcp", endpoint)
	if err != nil {
		logger.Fatalf("No socket listening at %s. Giving up: %s", endpoint, err)
		os.Exit(1)
	}
	return true
}

// Which checks to see if a given utility exists on the local machine
func Which(utility string) bool {
	_, err := exec.LookPath(utility)
	if err != nil {
		return false
	}
	return true
}

// ConsulBinaryCall runs an arbitrary command using the consul CLI utility
func ConsulBinaryCall(a, b string) string {
	out, err := exec.Command("consul", a, b).Output()
	if err != nil {
		message := fmt.Sprintf("There was an error querying consul. Giving up: %s", err)
		fmt.Println(message)
		os.Exit(1)
	}
	return string(out)
}

// GitBinaryCall runs an arbitrary command using the git CLI utility
func GitBinaryCall(cmd string) string {
	fullCommand := fmt.Sprintf("git %s", cmd)
	args := strings.Fields(cmd)
	logger.Debugf("Calling git with: `%s`", fullCommand)
	out, err := exec.Command("git", args...).CombinedOutput()
	if err != nil {
		message := fmt.Sprintf("There was an error running git. Giving up: %s", err)
		fmt.Println(message)
		os.Exit(1)
	}
	return string(out)
}

// DeleteEmpty removes the empty strings from a string array
func DeleteEmpty(s []string) []string {
	var r []string
	for _, str := range s {
		if str != "" {
			r = append(r, str)
		}
	}
	return r
}


backup.go
package main

import (
	"encoding/base64"
	"fmt"
	"github.com/hashicorp/consul/api"
	"os"
	"sort"
)

// ByCreateIndex sorts consul KV pairs
type ByCreateIndex api.KVPairs

func (a ByCreateIndex) Len() int           { return len(a) }
func (a ByCreateIndex) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a ByCreateIndex) Less(i, j int) bool { return a[i].CreateIndex < a[j].CreateIndex }

// Backup talks to Consul and starts reading KV pais and writes them to a file
func Backup(ipAddress string, token string, outfile string, exclusion []string, inclusion []string) {

	config := api.DefaultConfig()
	config.Address = ipAddress
	config.Token = token

	client, _ := api.NewClient(config)
	kv := client.KV()

	pairs, _, err := kv.List("/", nil)
	if err != nil {
		panic(err)
	}

	sort.Sort(ByCreateIndex(pairs))

	if len(pairs) == 0 {
		logger.Info("No KV pairs found. Nothing to do. Exiting.")
		os.Exit(0)
	} else {
		logger.Infof("Found %d KV pairs", len(pairs))
	}

	outstring := ""
	if len(exclusion) > 0 {
		for _, element := range pairs {
			if !StartsWith(exclusion, element.Key) {
				encodedValue := base64.StdEncoding.EncodeToString(element.Value)
				outstring += fmt.Sprintf("%s:%s\n", element.Key, encodedValue)
			}
		}
	} else if len(inclusion) > 0 {
		for _, element := range pairs {
			if StartsWith(inclusion, element.Key) {
				encodedValue := base64.StdEncoding.EncodeToString(element.Value)
				outstring += fmt.Sprintf("%s:%s\n", element.Key, encodedValue)
			}
		}
	} else {
		for _, element := range pairs {
			encodedValue := base64.StdEncoding.EncodeToString(element.Value)
			outstring += fmt.Sprintf("%s:%s\n", element.Key, encodedValue)
		}
	}

	file, err := os.Create(outfile)
	if err != nil {
		panic(err)
	}

	if _, err := file.Write([]byte(outstring)[:]); err != nil {
		panic(err)
	}

	logger.Infof("Wrote %d KV pairs to %s", len(pairs), outfile)
}

// BackupACLs talks to Consul and starts printing ACL rules currently in place
func BackupACLs(ipAddress string, token string, outfile string) {

	config := api.DefaultConfig()
	config.Address = ipAddress
	config.Token = token

	client, _ := api.NewClient(config)
	acl := client.ACL()

	tokens, _, err := acl.List(nil)
	if err != nil {
		panic(err)
	}

	outstring := ""
	for _, element := range tokens {
		outstring += fmt.Sprintf("====\nID: %s\nName: %s\nType: %s\nRules:\n%s\n", element.ID, element.Name, element.Type, element.Rules)
	}

	file, err := os.Create(outfile)
	if err != nil {
		panic(err)
	}

	if _, err := file.Write([]byte(outstring)[:]); err != nil {
		panic(err)
	}
}


restore.go
package main

import (
	"encoding/base64"
	"github.com/hashicorp/consul/api"
	"io/ioutil"
	"strings"
)

// Restore reads a file `infile` and saves the result to the Consul cluster at `ipAddress`
func Restore(ipAddress string, token string, infile string) {

	config := api.DefaultConfig()
	config.Address = ipAddress
	config.Token = token

	data, err := ioutil.ReadFile(infile)
	if err != nil {
		panic(err)
	}

	client, _ := api.NewClient(config)
	kv := client.KV()

	pairs := DeleteEmpty(strings.Split(string(data), "\n"))

	for _, element := range pairs {
		pair := strings.Split(element, ":")

		if len(pair) > 1 {
			val, decodeErr := base64.StdEncoding.DecodeString(pair[1])
			if decodeErr != nil {
				panic(decodeErr)
			}

			p := &api.KVPair{Key: pair[0], Value: val}
			_, err := kv.Put(p, nil)
			if err != nil {
				panic(err)
			}
		}
	}
	logger.Infof("Wrote %d KV pairs to Consul at %s from file %s", len(pairs), ipAddress, infile)
}


main.go
package main

import (
	"fmt"
	"github.com/docopt/docopt-go"
	"os"
	"strings"
)

func main() {

	usage := fmt.Sprintf(`Consul KV and ACL Backup and Sync tool.

Version: %s (Commit: %s)

Usage:
  consul-bak (backup|restore|aclbackup)
                [--leader-only]
                [--rpc-port RPCPORT]
                [--http-port HTTPPORT]
                [--address IP]
                [--include-prefix INPREFIX]...
                [--exclude-prefix EXPREFIX]...
                [--token TOKEN]
                [--no-prompt]
                <filename>
  consul-bak dumptree
                [--leader-only]
                [--rpc-port RPCPORT]
                [--http-port HTTPPORT]
                [--address IP]
                <pathname>
  consul-bak syncgit
                [--leader-only]
                [--rpc-port RPCPORT]
                [--http-port HTTPPORT]
                [--address IP]
                <git-url>
  consul-bak -h | --help
  consul-bak -v | --version

Options:
  -h, --help                         Show this screen.
  -v, --version                      Show version.
	--mode=MODE                        Set mode, can be one of backup,restore,syncgit,dumptree,aclbackup [default: backup]
  --leader-only                      Only run on consul leader.
  --rpc-port=RPCPORT                 RPC port [default: 8400].
  --http-port=HTTPPORT               HTTP endpoint port [default: 8500].
  --address=IP                       The HTTP endpoint of Consul [default: 127.0.0.1].
  --include-prefix=[INPREFIX]        Repeatable option for keys starting with prefix to include in the backup.
  --exclude-prefix=[EXPREFIX]        Repeatable option for keys starting with prefix to exclude from the backup.
  --token=TOKEN                      An ACL Token with proper permissions in Consul [default: ].
  --force                            Don't prompt, force overwrite.`, Version, GitCommit)

	arguments, _ := docopt.Parse(usage, nil, true, fmt.Sprintf("consul-bak %s (%s)", Version, GitCommit), false)
	SetupLogging()

	var (
		httpEndpoint = fmt.Sprintf("%s:%s", arguments["--address"], arguments["--http-port"])
		rpcEndpoint  = fmt.Sprintf("%s:%s", arguments["--address"], arguments["--rpc-port"])
		rpcOptString = fmt.Sprintf("-rpc-addr=%s", rpcEndpoint)
	)

	logger.Infof("Verifying HTTP endpoint: %s", httpEndpoint)
	CheckSocket(httpEndpoint)
	logger.Infof("Verifying RPC endpoint: %s", rpcEndpoint)
	CheckSocket(rpcEndpoint)

	if arguments["--leader-only"] == true {
		logger.Info("Running in leader only mode, only running on Consul leader.")
		// if consul client is not available we keep running
		if Which("consul") {
			var out = ConsulBinaryCall("info", rpcOptString)
			if strings.Contains(out, "leader = true") {
				logger.Info("Consul leader, continuing.")
			} else {
				logger.Error("Not a Consul leader, stopping.")
				os.Exit(1)
			}
		} else {
			logger.Error("Could not find `consul` utility. Is your $PATH setup properly?")
			os.Exit(1)
		}
	}

	if arguments["restore"].(bool) {
		logger.Info("Running in restore mode.")
		if (len(arguments["--exclude-prefix"].([]string)) > 0) || (len(arguments["--include-prefix"].([]string)) > 0) {
			logger.Error("--exclude-prefix, -x and --include-prefix, -n can be used only for backups")
			os.Exit(1)
		}
		if !arguments["--force"].(bool) {
			fmt.Printf("\nWarning! This will overwrite existing kv. Press [enter] to continue; CTL-C to exit")
			fmt.Scanln()
		}
		logger.Infof("Restoring KV from file: %s", arguments["<filename>"].(string))
		Restore(httpEndpoint, arguments["--token"].(string), arguments["<filename>"].(string))
	} else if arguments["backup"].(bool) {
		logger.Info("Running in backup mode.")
		if (len(arguments["--exclude-prefix"].([]string)) > 0) && (len(arguments["--include-prefix"].([]string)) > 0) {
			logger.Error("--exclude-prefix and --include-prefix cannot be used together")
			os.Exit(1)
		}
		if len(arguments["--exclude-prefix"].([]string)) > 0 {
			logger.Infof("excluding keys with prefix(es): %s", arguments["--exclude-prefix"].([]string))
		}
		if len(arguments["--include-prefix"].([]string)) > 0 {
			logger.Infof("including only keys with prefix(es): %s", arguments["--include-prefix"].([]string))
		}
		logger.Infof("KV store will be backed up to file: %s", arguments["<filename>"].(string))
		Backup(httpEndpoint, arguments["--token"].(string), arguments["<filename>"].(string), arguments["--exclude-prefix"].([]string), arguments["--include-prefix"].([]string))
	} else if arguments["aclbackup"].(bool) {
		logger.Infof("ACL Tokens will be backed up to file: %s", arguments["<filename>"].(string))
		BackupACLs(httpEndpoint, arguments["--token"].(string), arguments["<filename>"].(string))
	} else if arguments["syncgit"].(bool) {
		logger.Info("Running in sync mode.")
		Sync(httpEndpoint, arguments["--token"].(string), arguments["<git-url>"].(string))
	} else if arguments["dumptree"].(bool) {
		var path string = arguments["<pathname>"].(string)
		logger.Infof("Running in dump mode. Dumping to %s", path)
		Dump(httpEndpoint, arguments["--token"].(string), path)
	}
}
